File: test_resources/kotlin/enum_basic.kt
sealed interface MyEnumWithoutPositions {
    data class Option1(
        val p0: ULong,
    ) : MyEnumWithoutPositions

    data class Option2(
        val name: String,
    ) : MyEnumWithoutPositions

    data object Option3 : MyEnumWithoutPositions

    data object Option4 : MyEnumWithoutPositions

    companion object {
        fun createDefault(): MyEnumWithoutPositions = Option1(
            0UL,
        )

        fun readFromBuffers(reader: Long): MyEnumWithoutPositions {
            return when (val case = UInt.readFromBuffers(reader)) {
                0U -> {
                    val p0 = ULong.readFromBuffers(reader)

                    Option1(
                        p0 = p0,
                    )
                }
                1U -> {
                    val name = String.readFromBuffers(reader)

                    Option2(
                        name = name,
                    )
                }
                2U -> Option3
                3U -> Option4
                else -> throw IllegalArgumentException("Invalid enum value: $case")
            }
        }

        fun skipInBuffers(reader: Long, count: Int) {
            for (i in 0..<count) {
                readFromBuffers(reader)
            }
        }
    }

    fun writeToBuffers(writer: Long) {
        when (this) {
            is Option1 -> {
                p0.writeToBuffers(writer)
            }
            is Option2 -> {
                name.writeToBuffers(writer)
            }
            Option3 -> {}
            Option4 -> {}
        }
    }
}

File: test_resources/swift/enum_basic.swift
enum MyEnumWithoutPositions: TechPawsBuffersModel {
    case option1(
        UInt64
    )
    case option2(
        /* name */ String
    )
    case option3
    case option4

    static func createBuffersDefault() -> MyEnumWithoutPositions {
        return .option1(
            0
        )
    }

    static func readFromBuffers(
        _ bytesReader: TechPawsBuffersBytesReader
    ) -> Self {
        let caseValue = bytesReader.readUInt32()

        switch caseValue {
        case 0:
            let p0 = bytesReader.readUInt64()

            return .option1(
                p0
            )
        case 1:
            let name = String.readFromBuffers(bytesReader)

            return .option2(
                name
            )
        case 2:
            return .option3
        case 3:
            return .option4
        default:
            fatalError("Invalid value: \(caseValue)")
        }
    }

    static func skipInBuffers(
        _ bytesReader: TechPawsBuffersBytesReader,
        _ count: UInt64
    ) {
        for _ in 1...count {
            let caseValue = bytesReader.readUInt32()

            switch caseValue {
            case 0:
                let _ = bytesReader.readUInt64()
            case 1:
                let _ = String.readFromBuffers(bytesReader)
            case 2:
                continue
            case 3:
                continue
            default:
                fatalError("Invalid value: \(caseValue)")
            }
        }
    }

    func writeToBuffers(
        _ bytesWriter: TechPawsBuffersBytesWriter
    ) {

        switch self {
        case .option1(
            let p0
        ):
            bytesWriter.writeUInt32(0)
            bytesWriter.writeUInt64(p0)
        case .option2(
            let name
        ):
            bytesWriter.writeUInt32(1)
            name.writeToBuffers(bytesWriter)
        case .option3:
            bytesWriter.writeUInt32(2)
        case .option4:
            bytesWriter.writeUInt32(3)
        }
    }
}
